internal/cmd/plugin/logs/pretty/pretty.go

```diff
@@ -0,0 +1,242 @@
+/*
+Copyright The CloudNativePG Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pretty
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/cloudnative-pg/machinery/pkg/stringset"
+	"github.com/logrusorgru/aurora/v4"
+	"github.com/spf13/cobra"
+
+	"github.com/cloudnative-pg/cloudnative-pg/internal/cmd/plugin"
+)
+
+type prettyCmd struct {
+	loggers  *stringset.Data
+	pods     *stringset.Data
+	minLevel LogLevel
+}
+
+// NewCmd creates a new `kubectl cnpg logs pretty` command
+func NewCmd() *cobra.Command {
+	var loggers, pods []string
+	bf := prettyCmd{}
+
+	cmd := &cobra.Command{
+		Use:   "pretty",
+		Short: "Prettify CNPG logs",
+		Long:  "Reads CNPG logs from the standard input and pretty-print them for human consumption",
+		ValidArgsFunction: func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+			return plugin.CompleteClusters(cmd.Context(), args, toComplete), cobra.ShellCompDirectiveNoFileComp
+		},
+		Args: cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			bf.loggers = stringset.From(loggers)
+			bf.pods = stringset.From(pods)
+
+			recordChannel := make(chan logRecord)
+			recordGroupsChannel := make(chan []logRecord)
+
+			var wait sync.WaitGroup
+
+			wait.Add(1)
+			go func() {
+				bf.decode(cmd.Context(), os.Stdin, recordChannel)
+				wait.Done()
+			}()
+
+			wait.Add(1)
+			go func() {
+				bf.group(cmd.Context(), recordChannel, recordGroupsChannel)
+				wait.Done()
+			}()
+
+			wait.Add(1)
+			go func() {
+				bf.write(cmd.Context(), recordGroupsChannel, os.Stdout)
+				wait.Done()
+			}()
+
+			wait.Wait()
+			return nil
+		},
+	}
+
+	cmd.Flags().StringSliceVar(&loggers, "loggers", nil,
+		"The list of loggers to receive. Defaults to all.")
+	cmd.Flags().StringSliceVar(&pods, "pods", nil,
+		"The list of pods to receive from. Defaults to all.")
+	cmd.Flags().Var(&bf.minLevel, "min-level",
+		`Hides the messages whose log level is less important than the specified one.
+Should be empty or one of error, warning, info, debug, or trace.`)
+
+	return cmd
+}
+
+// decode progressively decode the logs
+func (bf *prettyCmd) decode(ctx context.Context, reader io.Reader, recordChannel chan<- logRecord) {
```

Comment:
Ah, lovely concurrent implementation.
Inspired by Rob Pike? https://go.dev/talks/2011/lex.slide#1

Reply:
No, it came out of my head! So it deserves to be reviewed carefully :)